### PR TITLE
Make yaml-cpp build with libc++

### PR DESCRIPTION
**Pull request by becker33**
When built with tests, yaml-cpp vendors gtest. gtest (at least how it is built within yaml-cpp) does not properly understand clang's libc++, and includes `cxxabi.h` headers that are only for use with libstdc++.

This PR does not fix the issue of yaml-cpp vendoring a test dependency. It merely presents a `tests` variant for yaml-cpp and expresses the conflict between that variant and `cxxflags=-stdlib=libc++`.

Because `cxxflags=-stdlib=libc++` is not satisfied by `cxxflags=-stdlib=libc++ -O3` or any other spec that includes the problem flag along with others, I've added handling in the flag handler to catch those situations and throw an error. It's not as good as a conflict, but it's a workaround until we think of something better.

This will allow users to build yaml-cpp with libc++, as long as they don't require the tests, and it presents usable, if not ideal, error handling.

### CODE REVIEW

File: var/spack/repos/builtin/packages/yaml-cpp/package.py [hunk 1]
```diff
@@ -32,6 +37,23 @@ class YamlCpp(CMakePackage):
...
+
+    def flag_handler(self, name, flags):
+        # We cannot catch all conflicts with the conflicts directive because
+        # the user can add arbitrary strings to the flags. Here we can at least
+        # fail early.
+        # We'll include cppflags in case users mistakenly put c++ flags there.
+        spec = self.spec
+        if name in ('cxxflags', 'cppflags') and spec.satisfies('+tests'):
+            if '-stdlib=libc++' in flags:
+                raise ConflictsInSpecError(spec,
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/yaml-cpp/package.py]: get rid of the global `yaml_cpp_tests_libcxx_error_msg` and just end this line with `[`.  Then everything below will be much less indented.

File: var/spack/repos/builtin/packages/yaml-cpp/package.py [hunk 2]
```diff
@@ -32,6 +37,23 @@ class YamlCpp(CMakePackage):
     conflicts('%intel@:11.1', when='@0.6.0:', msg="versions 0.6.0: require c++11 support")
     conflicts('%xl@:13.1', when='@0.6.0:', msg="versions 0.6.0: require c++11 support")
     conflicts('%xl_r@:13.1', when='@0.6.0:', msg="versions 0.6.0: require c++11 support")
+    conflicts('%clang cxxflags="-stdlib=libc++"', when='+tests',
+              msg=yaml_cpp_tests_libcxx_error_msg)
+
+    def flag_handler(self, name, flags):
+        # We cannot catch all conflicts with the conflicts directive because
+        # the user can add arbitrary strings to the flags. Here we can at least
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/yaml-cpp/package.py]: What do you mean "arbitrary strings"?  Not sure I understand this particular comment.

Comment (becker33) [reply var/spack/repos/builtin/packages/yaml-cpp/package.py]: @tgamblin The spec `cxxflags='-O3 -stdlib=libc++'` doesn't satisfy `cxxflags='-stdlib=libc++`, and the user can put in any string that's a valid flag for whichever compiler they're using. So we can't just enumerate the actual conflicts using the `conflicts` directive, so we need to do conflict management in the flag handler.

File: var/spack/repos/builtin/packages/yaml-cpp/package.py (hunk 1 above)
Comment (becker33) [reply var/spack/repos/builtin/packages/yaml-cpp/package.py]: @tgamblin The global makes it so we don't have the same message copied into two places (the other copy is in the `conflicts` directive. I can fix the indentation without changing that.

**Pull request merged**